src/lib.rs

```diff
@@ -124,96 +24,1138 @@ impl<T> RpcMessage for T where
+                    Sender::Tokio(tx) => {
+                        tx.send(value).await.map_err(|_| SendError::ReceiverClosed)
+                    }
+                    Sender::Boxed(sink) => sink.send(value).await.map_err(SendError::from),
+                }
+            }
+
+            pub async fn try_send(&mut self, value: T) -> std::result::Result<(), SendError> {
+                match self {
+                    Sender::Tokio(tx) => match tx.try_send(value) {
+                        Ok(()) => Ok(()),
+                        Err(tokio::sync::mpsc::error::TrySendError::Closed(_)) => {
+                            Err(SendError::ReceiverClosed)
+                        }
+                        Err(tokio::sync::mpsc::error::TrySendError::Full(_)) => Ok(()),
+                    },
+                    Sender::Boxed(sink) => {
+                        sink.try_send(value).await.map_err(SendError::from)?;
+                        Ok(())
+                    }
+                }
+            }
+        }
+
+        impl<T> crate::sealed::Sealed for Sender<T> {}
+        impl<T> crate::Sender for Sender<T> {}
+
+        pub enum Receiver<T> {
+            Tokio(tokio::sync::mpsc::Receiver<T>),
+            Boxed(Box<dyn BoxedReceiver<T>>),
+        }
+
+        impl<T: RpcMessage> Receiver<T> {
+            /// Receive a message
+            ///
+            /// Returns Ok(None) if the sender has been dropped or the remote end has
+            /// cleanly closed the connection.
+            ///
+            /// Returns an an io error if there was an error receiving the message.
+            pub async fn recv(&mut self) -> std::result::Result<Option<T>, RecvError> {
+                match self {
+                    Self::Tokio(rx) => Ok(rx.recv().await),
+                    Self::Boxed(rx) => Ok(rx.recv().await?),
+                }
+            }
+        }
+
+        impl<T> From<tokio::sync::mpsc::Receiver<T>> for Receiver<T> {
+            fn from(rx: tokio::sync::mpsc::Receiver<T>) -> Self {
+                Self::Tokio(rx)
+            }
+        }
+
+        impl<T> TryFrom<Receiver<T>> for tokio::sync::mpsc::Receiver<T> {
+            type Error = Receiver<T>;
+
+            fn try_from(value: Receiver<T>) -> Result<Self, Self::Error> {
+                match value {
+                    Receiver::Tokio(tx) => Ok(tx),
+                    Receiver::Boxed(_) => Err(value),
+                }
+            }
+        }
+
+        impl<T> Debug for Receiver<T> {
+            fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+                match self {
+                    Self::Tokio(_) => f.debug_tuple("Tokio").finish(),
+                    Self::Boxed(_) => f.debug_tuple("Boxed").finish(),
+                }
+            }
+        }
+
+        impl<T> crate::sealed::Sealed for Receiver<T> {}
+        impl<T> crate::Receiver for Receiver<T> {}
+    }
+
+    /// No channels, used when no communication is needed
```

Comment:
when does this happen?

Reply:
E.g. you have a message that has a oneshot return channel but does not need an update channel. The typical 1 -> 1 rpc pattern. The WithChannels struct still has `rx` and `tx` fields for simplicity, so you need something to put into the `rx` field. It's a zero size struct, so it does not cost you anything at runtime.

Reply:
got it 👍